### PR TITLE
:tada: Extend logic for JDBC connectors to provide additional properties in JSON schema

### LINKED ISSUE #7777: Extend SourceOperations logic for JDBC connectors to provide additional properties in JSON schema
## Current Behavior
Currently all jdbc connectors can provide only primitive json types 
( STRING, NUMBER, OBJECT, ARRAY, BOOLEAN, NULL)

and json schema looks like
```
{ "type": "string" },
```
```
{ "type": "number" }
```

## Expected Behavior
Update json schema creation logic to have ability to provide additional keywords to json type
Examples:
```
{
  "type": "string",
  "minLength": 2,
  "maxLength": 3
}
```
```
{
   "type": "string",
   "pattern": "^(\\([0-9]{3}\\))?[0-9]{3}-[0-9]{4}$"
}
```
```
{
   "type": "string",
   "format": "date-time"
}
```
```
{
  "type": "number",
  "minimum": 0,
  "exclusiveMaximum": 100
}
```

### PR DESCRIPTION
**Pull request by VitaliiMaltsev**
## What
Currently all jdbc connectors can provide only primitive json types
( STRING, NUMBER, OBJECT, ARRAY, BOOLEAN, NULL)

and json schema looks like

`{ "type": "string" },`
`{ "type": "number" }`

## How
JsonSchemaPrimitive class was updated with jsonSchemaTypeMap

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

**Review by etsybaev — APPROVED** [review]
LGTM. 
But please make sure you pass all PR's check-points (for example to match the PR's name convention) and it would be good to check some destinations to make sure nothing will be broken by such schema changing.

**Review by yurii-bidiuk — APPROVED** [review]

**Comment by VitaliiMaltsev** [comment]
> LGTM. But please make sure you pass all PR's check-points (for example to match the PR's name convention) and it would be good to check some destinations to make sure nothing will be broken by such schema changing.

currently destinations can not handle json schemas type with keywords and just skip the "format". So everything should be as previously

**Review by tuliren — COMMENTED** [review]
This is nice! Such a concise approach.

I have two questions / comments:
1. `JsonSchemaPrimitive` is widely used in so many connectors. It's surprising that this change does not affect any of them. Even though theoretically no destination will be affected, it's still safer to run integration tests on a few sources and destinations.
2. Have you thought about any alternative approach and compared them? E.g. replace `JsonSchemaPrimitive` with a class, which allows more flexible specifications. One concern about this approach is that to support more formats, there will be many more enums to add, and it does not handle things like `pattern`, `max`, etc. Using a class seems more flexible and extensible.

File: airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitive.java [hunk 1]
```diff
@@ -4,11 +4,28 @@
...
-  STRING,
-  NUMBER,
-  OBJECT,
-  ARRAY,
-  BOOLEAN,
-  NULL;
+
+  STRING_DATE(ImmutableMap.of("type", "string", "format", "date")),
+  STRING_TIME(ImmutableMap.of("type", "string", "format", "time")),
+  STRING_TIMESTAMP(ImmutableMap.of("type", "string", "format", "date-time")),
```
Comment (sherifnada) [inline airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitive.java:13]: i recommend changing this to `STRING_DATETIME` as it matches the json primitive it represents

File: airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitive.java [hunk 2]
```diff
@@ -4,11 +4,28 @@
 
 package io.airbyte.protocol.models;
 
+import com.google.common.collect.ImmutableMap;
+
 public enum JsonSchemaPrimitive {
```
Comment (sherifnada) [inline airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitive.java:9]: might make sense to call this `JsonSchemaType` as these no longer represent primitives

File: airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java [hunk 3]
```diff
@@ -317,9 +317,9 @@ public JsonSchemaPrimitive getType(final JDBCType bigQueryType) {
       case REAL -> JsonSchemaPrimitive.NUMBER;
       case NUMERIC, DECIMAL -> JsonSchemaPrimitive.NUMBER;
       case CHAR, NCHAR, NVARCHAR, VARCHAR, LONGVARCHAR -> JsonSchemaPrimitive.STRING;
-      case DATE -> JsonSchemaPrimitive.STRING;
-      case TIME -> JsonSchemaPrimitive.STRING;
-      case TIMESTAMP -> JsonSchemaPrimitive.STRING;
+      case DATE -> JsonSchemaPrimitive.STRING_DATE;
```
Comment (sherifnada) [inline airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java:320]: I think we need to be super careful about how MySQL/Postgres/MSSQL/Oracle handle timezones. For example, from the MySQL docs: 

> MySQL converts TIMESTAMP values from the current time zone to UTC for storage, and back from UTC to the current time zone for retrieval. (This does not occur for other types such as DATETIME.) By default, the current time zone for each connection is the server's time. The time zone can be set on a per-connection basis. As long as the time zone setting remains constant, you get back the same value you store. If you store a TIMESTAMP value, and then change the time zone and retrieve the value, the retrieved value is different from the value you stored. This occurs because the same time zone was not used for conversion in both directions. The current time zone is available as the value of the time_zone system variable. For more information, see Section 5.1.15, “MySQL Server Time Zone Support”.

So for MySQL to correctly express timestamps we would need to set the connection timezone to UTC whenever we're reading the date types in UTC (we should also have tests for this)

wdyt?

**Comment by tuliren** [comment]
@sherifnada, what do you think about changing `JsonSchemaPrimitive` from an `enum` to a `class`? It looks like enum won't be able to support more formats. I think eventually we need to switch to a class.

**Comment by sherifnada** [comment]
@tuliren good idea, we can even have the current enum be a set of public static constants in that class, but still make it flexible enough to have any valid fields in an object's definition

**Comment by tuliren** [comment]
Given the potential complicity of changing `JsonSchemaPrimitive` to a class, I am fine with merging this PR as is, and do that in a separate PR.

But it is still important to run integration tests on a few sources and destinations to make sure nothing is broken.

File: airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitive.java (hunk 1 above)
Comment (VitaliiMaltsev) [reply airbyte-protocol/models/src/main/java/io/airbyte/protocol/models/JsonSchemaPrimitive.java]: > i recommend changing this to `STRING_DATETIME` as it matches the json primitive it represents

done

**Comment by VitaliiMaltsev** [comment]
> Given the potential complicity of changing `JsonSchemaPrimitive` to a class, I am fine with merging this PR as is, and do that in a separate PR.
> 
> But it is still important to run integration tests on a few sources and destinations to make sure nothing is broken.

@tuliren i ran some tests for Oracle Source, CockroachDB Source, MySQL Source, Postgres Destination and MSSQL Destination. And also created a separate ticket for changing JsonSchemaPrimitive to a class https://github.com/airbytehq/airbyte/issues/7944. Can i merge this PR ?

File: airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java (hunk 3 above)
Comment (VitaliiMaltsev) [reply airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java:320]: good point. But currently jdbc destinations can not handle different date-time datatypes for databases and just write value as string

**Comment by tuliren** [comment]
@VitaliiMaltsev, yes. Thank you for running the tests, and creating the issue. Feel free to merge it.

File: airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceAcceptanceTest.java [hunk 4]
```diff
@@ -803,28 +803,32 @@ protected AirbyteCatalog getCatalog(final String defaultNamespace) {
...
             Field.of(COL_FIRST_NAME, JsonSchemaPrimitive.STRING),
             Field.of(COL_LAST_NAME, JsonSchemaPrimitive.STRING),
-            Field.of(COL_UPDATED_AT, JsonSchemaPrimitive.STRING))
+            Field.of(COL_UPDATED_AT, resolveJsonSchemaType()))
             .withSupportedSyncModes(Lists.newArrayList(SyncMode.FULL_REFRESH, SyncMode.INCREMENTAL))
             .withSourceDefinedPrimaryKey(
                 List.of(List.of(COL_FIRST_NAME), List.of(COL_LAST_NAME)))));
   }
 
+  private JsonSchemaPrimitive resolveJsonSchemaType() {
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jdbc/src/testFixtures/java/io/airbyte/integrations/source/jdbc/test/JdbcSourceAcceptanceTest.java:828]: shouldn't the oracle test instead override this method, instead of hardcoding a connector-specific edge case in the abstract test class? also, the name of this method should be clearer as `resolveJsonSchemaType` does not reflect it is only concerned with date types

File: airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java (hunk 3 above)
Comment (sherifnada) [reply airbyte-db/lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java:320]: @VitaliiMaltsev 

> currently jdbc destinations can not handle different date-time datatypes for databases and just write value as string

I believe this is inaccurate - [normalization writes appropriately typed date/time types depending on the value in the `format` field](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py#L500)

**Review by sherifnada — COMMENTED** [review]
@VitaliiMaltsev there's a few points we need to make sure are taken into account here: 
1. If I'm reading this correctly, the current tests don't cover all newly introduced types i.e: we must test that all 3 newly introduced types (date, time, datetime) are used appropriately. It's crucial that we maximize test coverage, otherwise customers will find issues before we do. 
2. JDBC connectors were not published in this PR, is this intended? in general whenever we make a change to a connector, we should publish it asap to make sure release notes are accurate and issues are caught quickly
3. I don't think we resolved the discussion around timezone. It is critical that we ensure (via comprehensive testing) there are no issues here. If I'm understanding this correctly, the MySQL example I mentioned could result in inaccurate date information being persisted in the destination, because depending on the timezone used in the JDBC connection, we might be formatting the output value incorrectly, which when writing normalized tables could result in the date format being misinterpreted as UTC when it is not (which would change the actual value of the date object)

**Comment by VitaliiMaltsev** [comment]
> @VitaliiMaltsev there's a few points we need to make sure are taken into account here:
> 
> 1. If I'm reading this correctly, the current tests don't cover all newly introduced types i.e: we must test that all 3 newly introduced types (date, time, datetime) are used appropriately. It's crucial that we maximize test coverage, otherwise customers will find issues before we do.
> 2. JDBC connectors were not published in this PR, is this intended? in general whenever we make a change to a connector, we should publish it asap to make sure release notes are accurate and issues are caught quickly
> 3. I don't think we resolved the discussion around timezone. It is critical that we ensure (via comprehensive testing) there are no issues here. If I'm understanding this correctly, the MySQL example I mentioned could result in inaccurate date information being persisted in the destination, because depending on the timezone used in the JDBC connection, we might be formatting the output value incorrectly, which when writing normalized tables could result in the date format being misinterpreted as UTC when it is not (which would change the actual value of the date object)
@sherifnada 
this PR has already been merged with the master branch. Is it necessary to revert the changes in this PR before these 3 points are resolved?

**Comment by VitaliiMaltsev** [comment]
@sherifnada please advise. I'm not sure - do we need to publish all jdbc connectors if changes were not in connectors code? 
Changes in this PR for airbyte-protocol and airbyte-db modules

**Comment by sherifnada** [comment]
`JdbcSourceOperations` (and others in this class) are leveraged by all JDBC connectors which means changing any such common class should trigger a publication of all JDBC connectors
